### DISCUSSION TITLE
Built in WakeOnLan option

### DISCUSSION THREAD

**Discussion by Gameniti** [1Remote/1Remote#365] [Ideas]
**Describe the solution you'd like**
I would like to have the option to enable WakeOnLan for some of my machines. I have some that only run when I need them and I currently have to start them manually. If you could add WOL as an option this would be fantastic.

== EDIT == (#448)

A built-in way to send a magic packet. Assigning a MAC-Address to a connection and either being able to manually send a WOL package with a button and/or automatically do it when trying to connect to an offline machine.

For the latter, it could be necessary to add a delay to the connection while the machine boots up.

**Comment by VShawn** [comment]
I have never used the WOL function (because I never shut down), I will later research how to implement waking up the computer through WOL and the possibility of integrating it into the APP. And I need to consider if the WOL feature is part of PRemoteM functionality.

In fact, you can still wake up your computer by writing a BAT script like this:

![image](https://user-images.githubusercontent.com/10143738/216755064-0bc66038-dd1d-4010-b6af-3830d709aea1.png)


=== EDIT === (#448)

As WOL is rarely used, I defined it as a script feature that users can implement on their own. 

You can ping the server and attempt WOL within the script: If the server is offline, you can set a sleep duration in the script after wol package is sent. After your WOL script concludes, 1RM will then establish the remote connection.

You can also open a VPN form the script, SEE http://1remote.org/usage/misc/pre-post-script-control/open-a-vpn-before-connect/.

Personally, I don't wish to integrate WOL into 1RM, especially since the 1RM already supports multiple remote addresses for a single server. Supporting WOL would require storing multiple MAC addresses for different network address, which would introduce additional complexity to the app and potentially result in more bugs.

**Comment by majkinetor** [comment]
Creting this feature seems simple enough, based on the [pwsh script](https://www.pdq.com/blog/wake-on-lan-wol-magic-packet-powershell) I am looking at now. We need remote computers MAC address though (maybe you get it when its ON, so you can use it later when its OFF). There is some info on it [here](https://stackoverflow.com/questions/12802888/get-a-machines-mac-address-on-the-local-network-from-its-ip-in-c-sharp). I would keep any potential solution cross platform though, because it will be easier some day to eventually port 1RM to linux.

Seems like not bad option to have. I always keep all my computers on though, but I could test this myself too.

**Comment by majkinetor** [comment]
I kinda like this feature and implementing it in the script is kind cumbersome - not because of the WOL but because you need to test when protocol service is up which depends on the protocol type.

BUT, 1RM also doesn't have this implementd, its availability test check for IP address, not protocol port so I agree with @VShawn that its a lot of work to be done correctly for probalby very low amount of users. I would like to have it for completness sake though one day, as ultimate remote manager deserves. The main problem seems to be to map MAC address to IP address which look possible based on link I posted [before](https://stackoverflow.com/questions/12802888/get-a-machines-mac-address-on-the-local-network-from-its-ip-in-c-sharp):

> I know such a task is achievable as this [Wake on LAN scanner](http://www.nirsoft.net/utils/wake_on_lan.html) software scans the local IP range and returns MAC address/Hostname of all on machines.

@Gameniti, why don't you try a script and let us know how that works? I already posted [link](https://www.pdq.com/blog/wake-on-lan-wol-magic-packet-powershell/) but here is something imediatelly usable:

1. https://github.com/MuraAtVwnet/PowerShell-WOL
2. https://github.com/timway/powershell-wol
3. https://github.com/PhilZ-cwm6/Send-WOL
4. https://github.com/bustacheeze/pwsh-wol
5. https://powershell.one/code/11.html

Scripts widely differ in length so you and I didn't test them. This should be more or less trivial.

Here are CLI tools so you can also create cmd.exe batch files:

- https://www.gammadyne.com/cmdline.htm#wol
- https://github.com/basildane/WakeOnLAN/releases/tag/2.12.4

**Reply by VShawn** [reply]
Then as a long-term goal, we need a lan scanner within 1RM and a protocol available detector

**Reply by majkinetor** [reply]
There is open source VB.NET tool I added above that seems to do this.
